tests/base.py

```diff
@@ -178,16 +178,15 @@ def write_processed_events(self, events):
 
         rows = []
         for event in events:
-            rows.append(self.dataset.row_from_processed_message(event))
+            rows.append(event)
 
         return self.write_rows(rows)
 
     def write_rows(self, rows):
         if not isinstance(rows, (list, tuple)):
             rows = [rows]
 
-        write_rows(
-            self.clickhouse,
-            dataset=self.dataset,
-            rows=rows,
+        NativeDriverBatchWriter(self.clickhouse).write(
```

Comment:
I am a little skeptic of explicitly referring to the specific concrete BatchWriter everywhere. Is there a specific reason for this instead of making all components depend on the abstraction (like we do for eventstream or tagstore in sentry) ?
That way will  also make us more rigorous about what we need to initialize the batchwriter. Right now, passing the clickhouse pool seems reasonable, but depending on the implementation makes it easier to add initialization fields that would not make sense for the HTTPBatchWriter.

Reply:
I'm not quite sure what you're suggesting with this comment — my (perhaps unclear) intention here was that everywhere that the `NativeDriverBatchWriter` is used, it could be substitutable with the abstract `BatchWriter` interface. Put another way, if we were able to fully commit to 3.7+ at this point, this could be rewritten and correctly type checked as:

```python
writer: BatchWriter = NativeDriverBatchWriter(self.clickhouse)
writer.write(self.dataset.get_schema(), rows)
```

Repeatedly instantiating the concrete class is just an artifact of this change being more or less a direct replacement for the `write_rows` method, since moving the point where the concrete `BatchWriter` is instantiated (e.g. a method on `DataSet` with the return type of `BatchWriter`) would also require the correct ClickHouse cluster object to be in scope at that point. Currently, there is no global ClickHouse pool, and sometimes the pool is instantiated with different options depending on the use (this is likely a smell in and of itself.)

Making that change seems (to me at least) like a broader change than makes sense to make here. I do think that we should make that change at some point, though (this also probably ties into your earlier comment about broader changes to the `DataSet` abstraction.)

Reply:
My suggestion was to indeed avoid depending on NativeDriverBatchWriter and instead depending on Batch writer. And to have a something like a factory method (till we have to be backward compatible) to depend on to get the right instance:

> def batch_writer(clickhouse):
>    return NativeDriverBatchWriter(clickhouse)

just as a way to avoid having the code depending on the implementation.
My reasoning for that is that, it would reduce the coupling between the code that needs to write and NativeDriverBatchWriter. 
This, on one side, makes it easier to introduce a different implementation (but this is minor, since the dependencies are only handful). 
More important is that it makes it clear that we should try to avoid passing initialization parameters to the batch writer that are specific to the Native implementation.

Anyway, if your plan is to eventually depend on the abstract class as soon as we commit to python 3.7 I am ok with this

Reply:
I think I understand now.

The issue that I see with that approach is that the constructor parameters of the `NativeDriverBatchWriter` and the upcoming `HTTPBatchWriter` are different — the native driver takes a `Client` (or pool) object, and the HTTP writer takes a host, port, and optional options (e.g. `{"load_balancing": "in_order"}`) map. This is intentional: note that there isn't a `__init__` method defined on [the abstract `BatchWriter` interface](https://github.com/getsentry/snuba/blob/595604c3c7ab02c296afe48e1b252e3cc193d4a3/snuba/writer.py#L9-L11) — just the `write` method. [Here is an example of the `HTTPBatchWriter` implementation](https://github.com/getsentry/snuba/blob/40acee6c1d4469f2c0815dbade1cc8cf32937721/snuba/writer.py#L41-L61) that will be the last significant chunk of this set of changes (other than #350.)

We could technically do what you suggest by making the factory parameterless (nullary function) since both of these objects could be instantiated from settings values. I'm a little hesitant to increase our dependence on global configuration, though. The approach that's currently implemented is at least a more lateral move.

Ultimately, I think whatever approach we take here will be temporary regardless. The writer instance shouldn't be being instantiated in the places it is now at all — it should be being defined elsewhere and be passed in (similar to how it's implemented in this change to `ConsumerWorker`) though I'm not sure where the correct initialization location actually is yet… probably somewhere closer to configuration though (perhaps on the `DataSet` class, but I'm not totally sure) and further away from the implementation — maybe this is sort of what you were getting at with [this comment](https://github.com/getsentry/snuba/pull/357#discussion_r302217173)?

Reply:
Yeah, the initialization problem was what I was concerned about. If we must initialize the class where we use it and the two implementations inherently require different parameters that are present only in the context where the class is initialized we cannot fully depend on the interface. Let's put it this way:
Do we ever expect to have a combination of NativeDriverBatchWriter and HttpDriverBatchWriter in production/dev/tests ? 
Or are we aiming at fully replacing one with the other (with possibly a transition time)? 
If we are aiming for the second, this is a non problem.

Reply:
Agreed we will eventually aim at depending on the abstraction instead of the implementation.